### PR TITLE
[prometheus-snmp-exporter] Add optional auth section

### LINKED ISSUE #3940: [prometheus-snmp-exporter] SNMPv3 auth
### Describe the bug a clear and concise description of what the bug is.

Hi,

I'm trying to get the SNMP Exporter to deliver metrics to Prometheus with SNMPv3 auth turned on. I am able to see data when I bring up https://snmp-exporter.milco.info/snmp?target=192.168.100.5&module=apcups&auth=snmpv3

However when I open Prometheus and check the Target Status it shows as down and the URL that Prometheus is checking is missing the `auth` parameter.

The relevant bits from the helm chart values.yaml file are below. My expectation is that the `auth` section in the params would be passed to Prometheus like the `module` and `target` and that it would be added to the url so that Prometheus can get the data. This is guess work on my part as I haven't been able to find much documentation around using the new configuration format for `auths`.

Any help is much appreciated.

Thanks,
Brian

### What's your helm version?

v3.12.1

### What's your kubectl version?

v1.28.2

### Which chart?

prometheus-snmp-exporter

### What's the chart version?

1.8.1

### What happened?

Prometheus returns the error: 

server returned HTTP status 400 Bad Request

### What you expected to happen?

Prometheus to show status up and collect the data from the exporter

### How to reproduce it?

`helm upgrade --install -n monitoring snmp-exporter prometheus-community/prometheus-snmp-exporter --version 1.8.1 -f ~/projects/prometheus/snmp.yaml`

Open: https://prom-admin.milco.info/prometheus/targets
find the snmp entries and view the status.

### Enter the changed values of values.yaml?

```
image:
  tag: "v0.24.1"

serviceMonitor:
  params:
    - name: apc-pdu-0
      target: 192.168.100.5
      auth:
        - snmpv3
      module:
        - apcups

config: |-
  auths:
    snmpv3:
      version: 3
      security_level: authPriv
      auth_protocol: SHA
      username: snmp_user
      password: auth_password
      priv_protocol: AES
      priv_password: priv_password
  modules:
    apcups:
      walk:
      - 1.3.6.1.4.1.318.1.1.1.12
      - 1.3.6.1.4.1.318.1.1.1.2
      - 1.3.6.1.4.1.318.1.1.1.3
      - 1.3.6.1.4.1.318.1.1.1.4
      [cut for brevity]
```

### Enter the command that you execute and failing/misfunctioning.

`helm upgrade --install -n monitoring snmp-exporter prometheus-community/prometheus-snmp-exporter --version 1.8.1 -f ~/projects/prometheus/snmp.yaml`

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by iPenguin**
#### What this PR does / why we need it

This PR passes the auth parameters available in newer versions of snmp-exporter from the chart to the application configuration.

#### Which issue this PR fixes

- fixes #3940

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Comment by iPenguin** [comment]
Hi @xiu, just checking to see if you have any feedback for this PR?

Thanks,
Brian

**Comment by walker-tom** [comment]
Thank you for your PR, similar changes have been implemented in #4086 but they didn't include a default value for the `ServiceMonitor` `auth`. If you'd like to update your PR to include this default value then I'll give it a review, thanks!

**Review by walker-tom — CHANGES_REQUESTED** [review]
Thank you for your PR, similar changes have been implemented in https://github.com/prometheus-community/helm-charts/pull/4086 but they didn't include a default value for the `ServiceMonitor` `auth`. If you'd like to update your PR to include this default value then I'll give it a review, thanks!

**Comment by iPenguin** [comment]
Hi @walker-tom I've fixed the merge conflict. Do you need me to do anything else?

Thanks,
Brian

**Comment by walker-tom** [comment]
> Hi @walker-tom I've fixed the merge conflict. Do you need me to do anything else?
> 
> Thanks, Brian

Hi Brian, thanks for updating your PR, could you update the Chart version in `Chart.yaml`, I think it got lost during the merge conflict. Thanks!

**Comment by iPenguin** [comment]
:facepalm: Thanks. I think that should fix it.

File: charts/prometheus-snmp-exporter/templates/servicemonitor.yaml [hunk 1]
```diff
@@ -22,6 +22,8 @@ spec:
     scrapeTimeout: {{ .scrapeTimeout | default $.Values.serviceMonitor.scrapeTimeout }}
     params:
       module: {{ .module | default $.Values.serviceMonitor.module }}
+      auth:
+      - {{ .auth | default $.Values.serviceMonitor.auth }}
```
Comment (walker-tom) [inline charts/prometheus-snmp-exporter/templates/servicemonitor.yaml:26]: Sorry only just noticed, there is already an `auth` field referenced on lines 29-31, could you remove those in favour of this please.

**Review by walker-tom — CHANGES_REQUESTED** [review]

**Review by walker-tom — APPROVED** [review]

**Comment by kevinastone** [comment]
This change seems to be broken in the default case where you don't specify an auth in the `params[*]`.  It's attempting to apply the default `$.Values.serviceMonitor.auth` (which is an array, defaulting to `[public_v2]`) to a single item in the array:

```
      auth:
      - {{ .auth | default $.Values.serviceMonitor.auth }}
```

 So it's ending up being:
 ```
    auth:
    - [public_v2]
```


Here's the error from my HelmRelease install:

```
ServiceMonitor.monitoring.coreos.com "snmp-exporter-prometheus-snmp-exporter-<target1>" is invalid: spec.endpoints[0].params.auth[0]: Invalid value: "array": spec.endpoints[0].params.auth[0] in body must be of type string: "array" && cannot patch "snmp-exporter-prometheus-snmp-exporter-<target2>" with kind ServiceMonitor: ServiceMonitor.monitoring.coreos.com "snmp-exporter-prometheus-snmp-exporter-<target2>" is invalid: spec.endpoints[0].params.auth[0]: Invalid value: "array": spec.endpoints[0].params.auth[0] in body must be of type string: "array"
````

**Comment by iPenguin** [comment]
> This change seems to be broken in the default case where you don't specify an auth in the `params[*]`. It's attempting to apply the default `$.Values.serviceMonitor.auth` (which is an array, defaulting to `[public_v2]`) to a single item in the array:

I apologize for the mistake. Thanks for fixing it.

**Comment by walker-tom** [comment]
@abctaylor You will need to create an `auths` configuration for your targets in your SNMP Exporter. As documented in [auth-split-migration.md](https://github.com/prometheus/snmp_exporter/blob/main/auth-split-migration.md) this is now separate to the `modules` configuration. You can see examples in [the generator file format](https://github.com/prometheus/snmp_exporter/tree/main/generator#file-format) for the settings available for SNMPv3. You then have at least two options for getting this configuration into your Container:

1. Create an `auths` configuration YAML file for the exporter in a `Secret` and mount that into your Container using `extraVolumes` and `extraVolumeMounts`, then pass in an argument `--config.file=/path/to/extra/auths/file` via the extra `extraArgs`. SNMP Exporter will then load up a this additional configuration
2. Add the `auths` configuration to your existing SNMP Exporter configuration file, set the relevant sensitive data as environment variables using `envFrom` and then reference those env vars in your `auths` section using `--config.expand-environment-variables` as mentioned [here](https://github.com/prometheus/snmp_exporter/blob/main/README.md#configuration)

Hope this helps.

**Pull request merged**